File: app/src/main/java/com/moufee/purduemenus/repository/FavoritesRepository.java
package com.moufee.purduemenus.repository;

import android.content.SharedPreferences;
import android.util.Log;

import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.Transformations;

import com.moufee.purduemenus.api.AuthenticatedAPITask;
import com.moufee.purduemenus.api.UpdateFavoritesTask;
import com.moufee.purduemenus.api.Webservice;
import com.moufee.purduemenus.db.FavoriteDao;
import com.moufee.purduemenus.menus.Favorite;
import com.moufee.purduemenus.menus.MenuItem;
import com.moufee.purduemenus.util.AppExecutors;

import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.UUID;

import javax.inject.Inject;

import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Response;

/**
 * Created by Ben on 13/08/2017.
 * Repository for Favorites
 */

public class FavoritesRepository {

    private FavoriteDao mFavoriteDao;
    private AppExecutors mAppExecutors;
    private Webservice mWebservice;
    private SharedPreferences mSharedPreferences;
    private static final String TAG = "FAVORITES_REPOSITORY";

    @Inject
    public FavoritesRepository(FavoriteDao favoriteDao, AppExecutors appExecutors, Webservice webservice, SharedPreferences sharedPreferences) {
        mFavoriteDao = favoriteDao;
        mAppExecutors = appExecutors;
        mWebservice = webservice;
        mSharedPreferences = sharedPreferences;
    }


    public LiveData<List<Favorite>> getFavorites() {
        return mFavoriteDao.loadAllFavorites();
    }

    public LiveData<Set<String>> getFavoriteIDSet() {
        return Transformations.map(mFavoriteDao.getFavoriteIDs(), favoriteIDs -> {
            if (favoriteIDs != null)
                return new HashSet<>(favoriteIDs);
            return new HashSet<>();

        });
    }

    public void updateFavoritesFromWeb(@Nullable String ticket) {
        mAppExecutors.networkIO().execute(new UpdateFavoritesTask(mSharedPreferences, mWebservice, mFavoriteDao).setTicket(ticket));
    }

    public void addFavorite(MenuItem item) {
        Favorite favorite = new Favorite(item.getName(), UUID.randomUUID().toString(), item.getId(), item.isVegetarian());
        mAppExecutors.diskIO().execute(() -> mFavoriteDao.insertFavorites(favorite));
        if (mSharedPreferences.getBoolean("logged_in", false))
            mAppExecutors.networkIO().execute(new AuthenticatedAPITask<ResponseBody>(mSharedPreferences) {
                @Override
                public Call<ResponseBody> getCall() {
                    return mWebservice.addFavorite(favorite);
                }

                @Override
                public void onSuccess(Response<ResponseBody> response) {
                    Log.d(TAG, "onSuccess: " + response);
                    updateFavoritesFromWeb(null);
                }
            });
    }

    public void removeFavorite(MenuItem item) {
        if (mSharedPreferences.getBoolean("logged_in", false))
            mAppExecutors.diskIO().execute(() -> {
                Favorite favorite = mFavoriteDao.getFavoriteByItemId(item.getId());
                Log.d(TAG, "removeFavorite: deleting favorite" + favorite + " " + favorite.favoriteId);
                mAppExecutors.networkIO().execute(new AuthenticatedAPITask<ResponseBody>(mSharedPreferences) {
                    @Override
                    public Call<ResponseBody> getCall() {
                        return mWebservice.deleteFavorite(favorite.favoriteId);
                    }

                    @Override
                    public void onSuccess(Response<ResponseBody> response) {
                        Log.d(TAG, "onSuccess: " + response);
                    }
                });
                mFavoriteDao.deleteByItemID(item.getId());
            });
        else
            mAppExecutors.diskIO().execute(() -> mFavoriteDao.deleteByItemID(item.getId()));

    }

    public void clearLocalFavorites() {
        mAppExecutors.diskIO().execute(() -> mFavoriteDao.deleteAll());
    }
}

File: app/src/main/java/com/moufee/purduemenus/repository/FavoritesRepository.kt
package com.moufee.purduemenus.repository

import com.moufee.purduemenus.api.models.RemoteFavorite
import com.moufee.purduemenus.repository.data.menus.Favorite
import com.moufee.purduemenus.repository.data.menus.MenuItem
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import timber.log.Timber
import java.util.*
import javax.inject.Inject
import javax.inject.Named

/**
 * Created by Ben on 13/08/2017.
 * Repository for Favorites
 */
class FavoritesRepository @Inject constructor(
    private val authDataSource: AuthDataSource,
    @Named(FavoritesDataSource.LOCAL) private val localDataSource: FavoritesDataSource,
    @Named(FavoritesDataSource.REMOTE) private val remoteDataSource: FavoritesDataSource,
) {
    val favorites: Flow<List<Favorite>>
        get() = localDataSource.observeFavorites()
    val favoriteIDSet: Flow<Set<String>>
        get() = favorites.map { favoriteList -> favoriteList.map { it.itemId }.toSet() }

    suspend fun updateFavoritesFromWeb(ticket: String?) {
        val remoteFavorites = remoteDataSource.getFavorites(ticket)
        localDataSource.addFavorites(remoteFavorites)
        val localFavorites = localDataSource.getFavorites()
        val localOnlyFavorites = localFavorites.subtract(remoteFavorites)
        for (favorite in localOnlyFavorites) {
            remoteDataSource.addFavorite(favorite)
        }
    }

    suspend fun addFavorite(item: MenuItem) {
        val favorite = Favorite(
            itemName = item.name,
            favoriteId = UUID.randomUUID().toString(),
            itemId = item.id,
            isVegetarian = item.isVegetarian
        )
        localDataSource.addFavorite(favorite)
        if (authDataSource.isLoggedIn()) {
            try {
                remoteDataSource.addFavorite(favorite)
                updateFavoritesFromWeb(null)

            } catch (t: Throwable) {
                Timber.e(t)
            }
        }
    }

    suspend fun removeFavorite(item: MenuItem) {
        localDataSource.getFavoriteByItemId(item.id)?.let { favorite ->
            if (authDataSource.isLoggedIn()) {
                remoteDataSource.removeFavorite(favorite)
            }
            localDataSource.removeFavorite(favorite)
        }
    }

    suspend fun clearLocalFavorites() {
        localDataSource.removeAllFavorites()
    }
}


fun Favorite.toApiFavorite() = RemoteFavorite(itemName, favoriteId, itemId, isVegetarian)

fun RemoteFavorite.toFavorite() = Favorite(ItemName, FavoriteId, ItemId, IsVegetarian)